jsf/subsystem/src/main/java/org/jboss/as/jsf/deployment/JSFDependencyProcessor.java

```diff
@@ -107,12 +107,16 @@ public void deploy(DeploymentPhaseContext phaseContext) throws DeploymentUnitPro
     public void undeploy(DeploymentUnit context) {
     }
 
-    private void addJSFAPI(String jsfVersion, ModuleSpecification moduleSpecification, ModuleLoader moduleLoader) {
-        if (jsfVersion.equals(JsfVersionMarker.WAR_BUNDLES_JSF_IMPL)) return;
+    private void addJSFAPI(String jsfVersion, ModuleSpecification moduleSpecification, ModuleLoader moduleLoader) throws DeploymentUnitProcessingException {
+        try{
+            if (jsfVersion.equals(JsfVersionMarker.WAR_BUNDLES_JSF_IMPL)) return;
+            ModuleIdentifier jsfModule = moduleIdFactory.getApiModId(jsfVersion);
+            ModuleDependency jsfAPI = new ModuleDependency(moduleLoader, jsfModule, false, false, false, false);
+            moduleSpecification.addSystemDependency(jsfAPI);
+        }catch (NullPointerException exception){
+            throw new DeploymentUnitProcessingException("JSF module not found for JSF version "+jsfVersion);
+        }
 
-        ModuleIdentifier jsfModule = moduleIdFactory.getApiModId(jsfVersion);
-        ModuleDependency jsfAPI = new ModuleDependency(moduleLoader, jsfModule, false, false, false, false);
-        moduleSpecification.addSystemDependency(jsfAPI);
     }
```

Comment:
I assume the NPE is coming from moduleSpecification.addSystemDependency(jsfAPI).

This addJSFAPI method is called from two places, L68 and L93. The L93 call has logic before it (L89-91) that checks for an invalid module slot and fails, which makes me think this is failing due to the L68 call. So does adding logic similar to L89-91 before the L68 call fix the problem?

Reply:
@bstansberry  It may fix. As per discussion in JBEAP-19759 we considered  throw an exception, the suspect was at 
L113 ModuleIdentifier jsfModule = moduleIdFactory.getApiModId(jsfVersion), not sure is it because of jsfVersion.  Currently we are not able to reproduce this issue but to be at safer side we thought handle exception. What do you suggest?

Reply:
It should be possible to write a test that leads to L68 being called, which is then a reproducible failure that can be fixed.

Since we already have logic that is meant to handle a bad jsfVersion before it leads to this kind of failure it should be applied in both relevant cases.  One reason for that is that logic already has an i18n failure message, and we don't want to add another message for essentially the same problem.

Catching an NPE and rethrowing is generally not good, because the NPE means some var was unexpectedly null. Better to prevent reaching that state by understanding how the var was null and dealing with that.

Reply:
@bstansberry  I've wrote testcase which lead to call L68. Still not able to reproduce NPE, as it is passing default JSF version "Mojarra-2.0". There is one way of getting NPE if moduleSpecification object is null.  Please suggest.

Reply:
AIUI, the value of JSFModuleIdFactory.defaultSlot is driven by the standalone.xml default-jsf-impl-slot attribute.

https://github.com/wildfly/wildfly/blob/master/jsf/subsystem/src/main/java/org/jboss/as/jsf/subsystem/JSFSubsystemAdd.java#L61
https://github.com/wildfly/wildfly/blob/master/jsf/subsystem/src/main/java/org/jboss/as/jsf/subsystem/JSFSubsystemAdd.java#L66
https://github.com/wildfly/wildfly/blob/master/jsf/subsystem/src/main/java/org/jboss/as/jsf/deployment/JSFVersionProcessor.java#L96

The bug is because the configuration has an invalid value for that attribute, so JSFModuleIdFactory.defaultSlot is a meaningless value.

This should result in ModuleIdFactory.getApiModId returning null, resulting in null getting passed into https://github.com/wildfly/wildfly/blob/master/jsf/subsystem/src/main/java/org/jboss/as/jsf/deployment/JSFDependencyProcessor.java#L133

That won't NPE in the code this PR is touching, but the bug report isn't about an NPE in that code. The bug report is about an NPE from org.jboss.modules.DependencySpec L637, and JSFDependencyProcessor.java#L133 passing null into the ModuleDependency constructor should result in DependencySpec L637 trying to process that null and an NPE.

ModuleIdFactory.getApiModId should return null for JSF version "Mojarra-2.0":

https://github.com/wildfly/wildfly/blob/master/jsf/subsystem/src/main/java/org/jboss/as/jsf/deployment/JSFModuleIdFactory.java#L170 calls 'computeSlot'.

computeSlot returns the (invalid) defaultSlot value:

https://github.com/wildfly/wildfly/blob/master/jsf/subsystem/src/main/java/org/jboss/as/jsf/deployment/JSFModuleIdFactory.java#L165

https://github.com/wildfly/wildfly/blob/master/jsf/subsystem/src/main/java/org/jboss/as/jsf/deployment/JSFModuleIdFactory.java#L170 does a map lookup in its 'apiIds' map using the invalid value, which will return null.  The 'apiIds' map is populated by scanning the module path so there's not going to be an entry in it for some invalid value.

Reply:
@bstansberry There is come thing wrong with this branch after I solved conflict. I will try to solve this tomorrow . If not then I will close this,  create new PR and tag this PR to it.   I have removed try-catch block and now code looks like below

```

@Override
    public void deploy(DeploymentPhaseContext phaseContext) throws DeploymentUnitProcessingException {
        moduleIdFactory = JSFModuleIdFactory.getInstance();
        final DeploymentUnit deploymentUnit = phaseContext.getDeploymentUnit();
        final DeploymentUnit tl = deploymentUnit.getParent() == null ? deploymentUnit : deploymentUnit.getParent();
        final ModuleSpecification moduleSpecification = deploymentUnit.getAttachment(Attachments.MODULE_SPECIFICATION);
        final ModuleLoader moduleLoader = Module.getBootModuleLoader();
        String jsfVersion = JsfVersionMarker.getVersion(tl).equals(JsfVersionMarker.NONE)? JSFModuleIdFactory.getInstance().getDefaultSlot() : JsfVersionMarker.getVersion(tl);
        String defaultJsfVersion = JSFModuleIdFactory.getInstance().getDefaultSlot();
        if(JsfVersionMarker.isJsfDisabled(deploymentUnit)) {
            if (jsfVersion.equals(defaultJsfVersion) && !moduleIdFactory.isValidJSFSlot(jsfVersion)) {
                throw JSFLogger.ROOT_LOGGER.invalidDefaultJSFImpl(defaultJsfVersion);
            }
            addJSFAPI(JsfVersionMarker.JSF_2_0, moduleSpecification, moduleLoader);
            return;
        }

```

Reply:
@bstansberry I have added latest changes. Please review.